### ISSUE TITLE
Netty NIO websocket thread stall under heavy load

### ISSUE THREAD

**Issue by plorenz** [netty/netty#8427]
### Expected behavior
When writing to a channel, the channel should eventually be emptied
### Actual behavior
We have a netty based application which is something like a gateway. It has N websockets writing to another application. When under heavy load we sometimes see websocket channels go into not writable state and not come out.

At a rough approximation, the code looks something like
```
public void forward(ByteBuffer message) {
  if (channel.isWritable()) {
    ByteBuf buf = Unpooled.wrappedBuffer(message);
    BinaryWebSocketFrame binaryMessage = new BinaryWebSocketFrame(buf);
    channel.writeAndFlush(message);
  } else {
     // reject message
  }
}
```
Notes
  * It seems to happen more easily with higher watermarks (1M/2M). 
  * It also happens using the default watermarks (32k/64k). 
  * I've also seen it recover, 1/2 hour to an hour after I stopped the simulator which was loading the application with incoming traffic. 
  * I've done heap dumps and can see that the ChanneOutboundBuffer is not writable, has X totalPendingSize greater than the high watermark, and this number doesn't change in heap dumps taken minutes later.  
  * For some reason we often see all the channel but 1 go into this state. The last channel goes in/out of writable repeatedly but doesn't seem to stall. Not sure why that is.
  * The netty IO thread associated with this Channel seems never to be busy. - see thread dump below

```
"NettyClient-NIO-2" #14 daemon prio=5 os_prio=0 tid=0x00007fbfb8a39000 nid=0x4626 runnable [0x00007fbf85fdd000]
   java.lang.Thread.State: RUNNABLE
        at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
        at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
        at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
        at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
        - locked <0x0000000080206790> (a io.netty.channel.nio.SelectedSelectionKeySet)
        - locked <0x0000000080207868> (a java.util.Collections$UnmodifiableSet)
        - locked <0x00000000802066e8> (a sun.nio.ch.EPollSelectorImpl)
        at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
        at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:62)
        at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:753)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:408)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at java.lang.Thread.run(Thread.java:748)
```
  * We're writing to the channel from outside the netty threads. I tried adding a flush in the pipeline when `channelWritabilityChanged` fired and writable was false, but it didn't seem to solve the problem (and also seems like it shouldn't be needed, since we're already calling writeAndFlush. 
  * We had a bug where we were writing to non-writable channels, but repeatedly calling writeAndFlush didn't seem to unstall it. The ChannelOutboundBuffer just filled up more. 

Aside from the minimal code to reproduce this, which I'm working on, is there any other information I can provide? 

### Minimal yet complete reproducer code (or URL to code)
Working on writing minimal code to reproduce this, but haven't gotten there yet. 

### Netty version
4.1.30

### JVM version (e.g. `java -version`)
```java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)
```
### OS version (e.g. `uname -a`)
Tested on Linux and windows
`Linux plorenz-lt 4.15.0-36-generic #39-Ubuntu SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
`

**Comment by plorenz** [comment]
Here's an example of the websocket channels becoming writable again after having been unwritable and stopping the flow of data.

```
12:14:38.077 [NettyClient-NIO-5] INFO  NettyChannelHandler - Channel is now writable after not being writable for 2165896 milliseconds: [id: 0x528a21ba, L:/127.0.0.1:19754 - R:localhost/127.0.0.1:8090]
12:15:34.097 [NettyClient-NIO-1] INFO  NettyChannelHandler - Channel is now writable after not being writable for 2186286 milliseconds: [id: 0x21c8dad1, L:/127.0.0.1:19750 - R:localhost/127.0.0.1:8090]
12:22:33.744 [NettyClient-NIO-2] INFO  NettyChannelHandler - Channel is now writable after not being writable for 2167037 milliseconds: [id: 0x25b75d1c, L:/127.0.0.1:19752 - R:localhost/127.0.0.1:8090]
12:26:05.956 [NettyClient-NIO-3] INFO  NettyChannelHandler - Channel is now writable after not being writable for 2166218 milliseconds: [id: 0xb1ba1b9d, L:/127.0.0.1:19756 - R:localhost/127.0.0.1:8090]
12:28:22.068 [NettyClient-NIO-4] INFO  NettyChannelHandler - Channel is now writable after not being writable for 2166172 milliseconds: [id: 0x1c34d9cd, L:/127.0.0.1:19748 - R:localhost/127.0.0.1:8090]
```
The times are all very similar, around 2166000ms, which is around 36 minutes.

**Comment by plorenz** [comment]
Adding `((NioUnsafe)currentChannel.unsafe()).forceFlush();` didn't have any effect.
I also tried using EPoll insteadof Nio, and also the same issue there.

**Comment by normanmaurer** [comment]
Thats very strange...  So you say that you can basically see that there are still pending bytes in the ChannelOutboundBuffer when it happens ? How many of these are "flushed" ?

**Comment by normanmaurer** [comment]
@plorenz ^^

**Comment by plorenz** [comment]
Hi @normanmaurer, apologies for the slow response. Couple of discoveries since I last posted.

1. The minimal test case I put together to try and reproduce this doesn't fail in the way described above. I've tried to complicate it in various ways to get it closer to the case that does fail, but haven't found it yet.
2. The case where we were seeing this in production turned out to a misinterpretation of log output. Because of a bug, we were continuing to push data to netty channels that were over the watermark. Once they were all over the watermark, we paused. This made it look like all the channels except one went into not-writable and never came out.

Since I can only get this to happen locally and it I can't get a minimal test case, I'm going to assume for now that this is not a Netty issue. If I can track it down further and it ends up being something netty related, I'll reopen the bug with a minimal reproducer.
